### PR TITLE
Scala Lang website redesign

### PR DESCRIPTION
**Pull request by dominv**
This PR contains a changes related to design, layouts and responsive support. We've been coordinating with Heather and the rest of the Scala Center team during the different stages of this work, receiving feedback from their end from wireframe design to the actual implementation in Jekyll.

Here is a summary of the main changes code-wise:

- All static content (i.e.: courses, trainings, events...) was changed from categorized blogs to Jekyll custom collections. In the previous version, categorized blogs were used to hold this data (as empty posts with a YAML front matter). As part of this work, the blog and the news section were condensed into one categorized blog, and we took this chance to move that content to independent custom collections as it made more sense and simplified the blog listing code. This involves moving the previous content to different folders (Jekyll custom collections live in folders named as the collection but with an underscore prefix), that's the main reason this PR has so many changed files: a large part of them are simply moved to different places.

- As the main part of the work was related to design, new layouts were created (and the old were removed to avoid confussion). Part of the static information that was hardcoded in some of these was moved to either custom collections (i.e.: online courses), YAML variables inside each page markdown file, or YAML data files (under the _data directory) where it did make sense, to help making this data easier to handle and update.

@heathermiller @MasseGuillaume could you please review it?

Thanks!

### CODE REVIEW

**Comment by SethTisue** [comment]
would it be possible to have this running live somewhere before it's merged, so interested reviewers can see it? (I realize it's possible for people to build the site locally to preview it, but that involves some Ruby setup and so forth — it's a bit of barrier.)

**Comment by SethTisue** [comment]
The number of commits and the number of files changed is a bit overwhelming. Could you summarize the major changes? Also, can you call reviewers' attentions to changes you're especially interested in getting feedback on?

**Comment by tpolecat** [comment]
In case anyone else gets shut down by not having Ruby 2.4, I managed by random experimentation to get this to build on Ruby 2.0 by first changing the version to `2.0.0` in `Gemfile` and then doing

```
rm Gemfile.lock
bundle install
bundle exec jekyll serve
```

It seems to work but I have no idea what else I might have done. Confidence high!

File: README.md [hunk 1]
```diff
@@ -97,7 +97,34 @@ The least error-prone way to link between documents, to link to local images, or
 
 Here, `{{ site.baseurl }}` is a site-wide variable that represents the root directory of the static site. So, to display the Scala logo image, located in `img/scala-logo.png`, one must simply write: `![Img alt text]({{ site.baseurl }}/resources/img/scala-logo.png)`
 
+## Permalinks
 
+In this new version of the scala-lang site we've tried to follow a `pretty permalink` style, so that any generated page will have an permalink finishing in a slash character (`/`). This will tell Jekyll to build that particular page as an `index.html` inside a folder with a name as specified in the provided permalink. i.e.: if a page has a permalink as follows:
+
+`permalink: /what-is-scala/`
+
+This will tell Jekyll to create a `what-is-scala` folder, with an `index.html` file inside. Links to this page will refer to the `{{site.baseurl}}/what-is-scala/`.
```
Comment (julienrf) [inline README.md]: Do previous links still work?

**Comment by heathermiller** [comment]
Hey everybody, FWIW there is a URL where you can see these changes generated: https://scala-lang-site.herokuapp.com/

There are lots of links we'll have to update, especially community-related links. Their changes were focused most on fixing layout, not content. The work of 47 degrees was just focused on making the site responsive and then updating the layouts on a few inner pages that needed some attention such as blog pages, the blog index page, etc. (When I made the original site, I didn't give these inner pages a lot of attention, layout-wise/CSS-wise)

**Comment by heathermiller** [comment]
Also, since this was a contracting job (the Scala Center pays hourly), I have already reviewed several intermediate steps of these changes. Because it's a contracting job, we couldn't afford to pay for many rounds of fielding and addressing community feedback on each iteration – I asked that we go for regular rounds of community feedback in the beginning, but it was suggested that it would quickly blow up the budget. So instead, I focused on (1) making many layout changes that we've been long discussing on Gitter and in GitHub issues (such as the responsiveness issue, and including IDEs and Scala.js on the front page of scala-lang, while (2) trying to streamline 47 Degrees' engineers' time so this stayed within budget.

**Comment by lrytz** [comment]
@heathermiller thanks for these explanations! While reviewing this, I think many people will have suggestions not directly related to the changes in this PR, like content or site structure. Can you remind us what's the preferred way to raise those / check for existing issues?

**Comment by heathermiller** [comment]
@lrytz Yeah, so there are lots of content updates that need to be made to this that I know of. I was hoping to make these on the same branch. 

Stuff like:

- [ ] fixing links to Gitter channels (there's a random list right now)
- [ ] removing the Scala Users Discourse channel links (we want to move eventually, so we wanted this in the layout, but those links shouldn't be there right now)
- [ ] adding all of the new getting started documentation by @travis032654 (there are currently bogus links on the download page that should go to those pieces of documentation)
- [ ] lots of updates to wording on the Download page
- [ ] syncing with the Scala team because I imagine that the updated Download page will affect make-release-notes etc.

etc... 

Since the 47 Degrees contracting job is over, these are of course content-related changes that we'll have to make. So I was planning on making many of them myself. I think the best thing to do is:

1. to discuss and collect feedback here, 
2. collect concrete, actionable blockers in a meta-ticket of todos that have to be completed (on this branch) before this is merged and made live.

**Comment by jdesiloniz** [comment]
@heathermiller @tpolecat @SethTisue @julienrf As Heather just commented, we did a lot of changes related to design, layouts and responsive support. We've been coordinating with Heather and the rest of the Scala Center team during the different stages of this work, receiving feedback from their end from wireframe design to the actual implementation in Jekyll.

I know that the sheer amount of changes in this PR can be daunting, so I'm going to try to summarize the main changes code-wise:

- All static content (i.e.: courses, trainings, events...) was changed from categorized blogs to Jekyll custom collections. In the previous version, categorized blogs were used to hold this data (as empty posts with a YAML front matter). As part of this work, the blog and the news section were condensed into one categorized blog, and we took this chance to move that content to independent custom collections as it made more sense and simplified the blog listing code. This involves moving the previous content to different folders (Jekyll custom collections live in folders named as the collection but with an underscore prefix), that's the main reason this PR has so many changed files: a large part of them are simply moved to different places.

- As the main part of the work was related to design, new layouts were created (and the old were removed to avoid confussion). Part of the static information that was hardcoded in some of these was moved to either custom collections (i.e.: online courses), YAML variables inside each page markdown file, or YAML data files (under the `_data` directory) where it did make sense, to help making this data easier to handle and update. 


Regarding SethTisue's suggestion about having the changes live: Heather already provided the URL of our Heroku deployment for these changes. We have been using Heroku and Travis to automatically deploy each PR so they could be reviewed by Heather and the team. It may make sense to set up something like this for the main repository if it makes sense to you guys. @dominv could help with that.

I think that's mostly it. Since I was the main developer of these changes I can help you with any doubts that may arise, and also @dominv who is now in charge of the development of this version of the site.

Thanks for your feedback!

**Comment by lrytz** [comment]
@heathermiller that sounds good. To echo what I heard: given the size of this PR, you will focus on a minimal set of changes that need to be done before this can be go live.

For suggestions outside that scope, should they be issues in this repo, like https://github.com/scala/scala-lang/issues/536?

**Comment by heathermiller** [comment]
Ah yeah, good point @lrytz. 

Bigger content-related changes should probably be new issues rather than part of the meta-ticket we'll eventually create with remaining todos. 

Just to comment specifically on #536... #536 is actually addressed by a combination of layout changes in this PR (on the download page) and the getting started documentation that @travis032654 wrote (but has yet to PR). Just a heads up. I assume that @travis032654 will likely PR these documents based off of this branch?

File: _data/common.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# SNS URLs
+githubUrl: https://github.com/scala/scala
+twitterUrl: https://twitter.com/scala_lang
+scaladexUrl: https://scaladex.scala-lang.org
```
Comment (MasseGuillaume) [inline _data/common.yml]: `https://index.scala-lang.org`

File: resources/js/functions.js [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        if (query == "") {
+            hideSuggestions()
+        } else {
+            showSuggestions();
+        }
+    },
+    transformResult: function(response) {
+        return {
+            suggestions: $.map(response, function(dataItem) {
+                return { value: dataItem.repository, data: 'https://scaladex.scala-lang.org/' + dataItem.organization + "/" + dataItem.repository };
```
Comment (MasseGuillaume) [inline resources/js/functions.js]: https://index.scala-lang.org

File: resources/js/functions.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        $("#result-container").show();
+      } else {
+        $("#result-container").hide();
+      }
+    });
+  }
+});
+
+// Scala in the browser
+$(document).ready(function() {
```
Comment (MasseGuillaume) [inline resources/js/functions.js]: Here is the javascript api to embed Scastie. I think there still some work to do to get the proper CSS
https://scastie.scala-lang.org/embedded-demo

Comment (heathermiller) [inline resources/js/functions.js]: This isn't necessary for this PR to be merged. Before we start worrying about embedding + scastie, let's first get scastie out of beta. Embedding scastie on scala-lang is something we can always do in the future if scastie is ever ready for it. But for now, this shouldn't block this PR.

**Comment by travis032654** [comment]
@heathermiller here is the pull request for the Getting Started tutorials: https://github.com/scalacenter/scala-lang/pull/1

**Comment by SethTisue** [comment]
This looks great overall! Excellent job.

So far I can only think of one change that I would request _before_ this is merged, which is that ScalaIDE should be included prominently in the IDE section.

I think https://scala-lang-site.herokuapp.com/download/ probably needs some further changes, but it could happen after merge.

**Comment by SethTisue** [comment]
> syncing with the Scala team because I imagine that the updated Download page will affect make-release-notes etc

I opened https://github.com/scala/scala-dev/issues/319 on it, so we won't forget, and and notified Adriaan.

**Comment by martijnhoekstra** [comment]
Was it intended to move the EPFL Red Dot into Lake Geneva?

**Comment by som-snytt** [comment]
It's a status indicator that everything is going swimmingly.

**Comment by dominv** [comment]
@martijnhoekstra @som-snytt The EPFL position marker is already fixed now. Thanks!

**Comment by MasseGuillaume** [comment]
Is it possible to redeploy here: https://scala-lang-site.herokuapp.com/ ?

**Comment by MasseGuillaume** [comment]
Eclipse should be labeled Scala IDE and point to http://scala-ide.org/

**Comment by MasseGuillaume** [comment]
I think the front page Run on section should also include Scala-Native and Scala-Android.

Since not all those target has the same level of maturity a visual cue should be added (different size, experimental ribbon, or other).

The same idea applies for the IDE section, they don't have the same level of quality.

**Comment by travissarles** [comment]
@dominv are you using any custom sass? I'm trying to figure out how the css got generated but I don't see any sass files besides what's in themes.

**Comment by jarrodwb** [comment]
Is there any word on this PR? I am interested in contributing to this site again but I feel a bit blocked. It would be great to get this PR finished up and merged in.

Is there anything I can do to help make this happen?

**Comment by heathermiller** [comment]
So @travissarles just went on vacation... But on Friday I believe @fsalvi had generated Travis's most recent PR and staged it somewhere on scala-lang so we can make sure everything works... 

@fsalvi is that link available somewhere? (Also is it available to people outside of EPFL?)

**Comment by fsalvi** [comment]
Yes, it's here:
http://www-dev.scala-lang.org/

**Comment by heathermiller** [comment]
Thanks @fsalvi! So, indeed what @travissarles has been working on the past couple of weeks is broken... Not sure what to do now that he's on vacation. The relevant PR is: https://github.com/scala/scala-lang/pull/667

**Comment by jarrodwb** [comment]
I don't mind waiting for @travissarles to come back from vacation. I will be around then too. :-) In the meantime, I will poke around and see what I can do.

**Comment by heathermiller** [comment]
(@jarrodwb: moving conversation to #667 since any bugfixes should go there)

**Comment by travissarles** [comment]
@dominv do you know why "online courses" don't show up when I build from your master branch? It looks good on Heroku for some reason.

**Comment by dominv** [comment]
@travissarles The "online courses" don't show up because the courses' `end-date`s have already passed. (They appear on Heroku because the site was built before those end-dates). Updating the courses dates in their own md files it should work properly. 

I've also detected that, for some reason, the courses' links are pointing to their own generated html files instead of linking to the specialization's url specified in the field with the same name. I've checked that changing this `url` field for something like `course-url` it works well. 

Please, let me know if I can be of any further help.

**Comment by jarrodwb** [comment]
:champagne: :clinking_glasses:
